velox/exec/AssignUniqueId.cpp

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "velox/exec/AssignUniqueId.h"
+
+#include <utility>
+
+namespace facebook::velox::exec {
+
+AssignUniqueId::AssignUniqueId(
+    int32_t operatorId,
+    DriverCtx* driverCtx,
+    const std::shared_ptr<const core::AssignUniqueIdNode>& planNode,
+    int32_t uniqueTaskId,
+    std::shared_ptr<std::atomic_int64_t> rowIdPool)
+    : Operator(
+          driverCtx,
+          planNode->outputType(),
+          operatorId,
+          planNode->id(),
+          "AssignUniqueId") {
+  isIdentityProjection_ = true;
+  rowIdPool_ = std::move(rowIdPool);
+
+  uniqueValueMask_ = ((int64_t)uniqueTaskId) << 40;
+  rowIdCounter_ = rowIdPool_->fetch_add(rowIdsPerRequest_);
+  maxRowIdCounterValue_ =
+      std::min(rowIdCounter_ + rowIdsPerRequest_, maxRowId_);
+}
+
+void AssignUniqueId::addInput(RowVectorPtr input) {
+  auto numInput = input->size();
+  VELOX_CHECK_NE(
+      numInput, 0, "AssignUniqueId::addInput received empty set of rows")
+  input_ = std::move(input);
+}
+
+RowVectorPtr AssignUniqueId::getOutput() {
+  if (input_ == nullptr) {
+    return nullptr;
+  }
+
+  BufferPtr nulls;
+
+  if (input_->nulls() != nullptr) {
```

Comment:
What is this logic for? Top-level rows can never be nulls. Individual columns may have null values, but top-level row cannot.